### ISSUE TITLE
bun install returns exit code 0 when it fails to install dependency

### ISSUE THREAD

**Issue by karl-run** [oven-sh/bun#4200]
### What version of Bun is running?

v0.7.3

### What platform is your computer?

github action with ubuntu-latest, is platform independent

### What steps can reproduce the bug?

Have a dependency in your lockfile that gives 401 on fetch, install dependencies:

Github action with `bun install --frozen-lockfile`:
![image](https://github.com/oven-sh/bun/assets/1507032/6c13800e-d7d0-4958-9fb8-a4160f5ae03a)

Locally with `bun install`
![image](https://github.com/oven-sh/bun/assets/1507032/a167ef42-66c9-4d47-8abe-d0ca28b37c84)

### What is the expected behavior?

That bun install returns exit code 1

**Comment by Hanaasagi** [comment]
encountered similar issue in bun 1.0.2. Some tarball downloaded failed, but it still displayed "install success" and the process's exit code was 0.


![2023-09-16_15-33](https://github.com/oven-sh/bun/assets/9482395/fdc6d679-ef40-4999-9b9a-837efac268b5)

**Comment by Jarred-Sumner** [comment]
> encountered similar issue in bun 1.0.2. Some tarball downloaded failed, but it still displayed "install success" and the process's exit code was 0.
> 
> ![2023-09-16_15-33](https://user-images.githubusercontent.com/9482395/268433080-fdc6d679-ef40-4999-9b9a-837efac268b5.png)

Can you check if the package exists in node_modules? It's supposed to automatically retry a few times, so I wonder if it failed once and then retried?

**Comment by Hanaasagi** [comment]
![2023-09-16_15-44](https://github.com/oven-sh/bun/assets/9482395/3db80f25-0644-41f3-be4e-d3b3ecbc6dd1)

`fast-decode-uri-component` is not here.

**Comment by Hanaasagi** [comment]
### Repro


```Python
# addon.py
from mitmproxy import http


def request(flow: http.HTTPFlow):
    if flow.request.path.endswith("moment-2.29.4.tgz"):
        flow.response = http.Response.make(
            502,  # 401 or Connection error ....
        )

```

Start mitmproxy

```
mitmproxy -p 1087 --ssl-insecure -s addon.py
```

Install package (make sure you have no local cache)

```
http_proxy=http://127.0.0.1:1087 https_proxy=http://127.0.0.1:1087 bun add moment
```
![2023-09-16_18-37](https://github.com/oven-sh/bun/assets/9482395/e4cb242e-0e47-42f4-9188-83ebb491e8c6)

**Comment by amphetamarina** [comment]
https://github.com/Hanaasagi/bun/commit/7055980aec71f7cd7a022b48eefb0e45af5a75a5

@Hanaasagi does it fix it? could you create PR for it? Thanks :)

**Comment by Hanaasagi** [comment]
@marinacabs 
Unfortunately, this commit didn't fix it correctly. There are other cases that still need to be addressed.

**Comment by palmithor** [comment]
Having this fixed would be great. 

It's pretty hard to debug for example GitHub actions workflow when this happens. 

In my case I just got a type error when building and didn't even think about checking the install output as it was successful.

I am able to reproduce on 1.0.17

**Comment by tgrushka** [comment]
There should be a flag (perhaps default behavior) for it to exit immediately on a 401 error instead of continuing to run. I don't want my Dockerfile to finish building if there's a 401 error because it didn't pick up the token environment variable. Right now there seems no way to trap this without waiting for the whole install to finish.

**Comment by tgrushka** [comment]
This is absurd, but works for now in Dockerfile:
```
RUN set -e; \
    bun install --production --frozen-lockfile 2>&1 | tee /tmp/bun.log; \
    grep -q ' - 401' /tmp/bun.log && { echo 'Error: Found "- 401" in output'; exit 1; } || true
```

**Comment by elyobo** [comment]
Still a problem with `1.0.26+c75e768a6` (see https://github.com/oven-sh/bun/issues/8863); in that case I ran into in GitHub Actions installing a private package and I had not configured authentication yet and no caching was involved.

**Comment by LucyEgan** [comment]
Still a problem in `1.1.12` causing our build to not fail at the right point when a internal repository wasn't available

```
6.135 error: FailedToOpenSocket downloading tarball @lupimedia/rollup-plugin-scss-template-literals@1.0.3                                                                                                                                                    
6.136                                                                                                                                                                                                                                                        
6.136 error: ConnectionRefused downloading tarball @lupimedia/rollup-plugin-expose-scoped-variables@1.1.1
6.136 
6.136 error: FailedToOpenSocket downloading tarball @lupimedia/firebase-polyfill@1.0.31
```

**Comment by dylan-conway** [comment]
This was fixed in #11828 and will be included in the bun `v1.1.14` release later today.
